core/src/main/java/com/linecorp/armeria/common/stream/DefaultStreamMessage.java

```diff
@@ -131,19 +129,28 @@ SubscriptionImpl subscribe(SubscriptionImpl subscription) {
 
     @Override
     public void abort() {
+        abort0(AbortedStreamException::get);
+    }
+
+    @Override
+    public void abort(Throwable cause) {
+        requireNonNull(cause, "cause");
+        abort(() -> cause);
+    }
+
+    @Override
+    public void abort(Supplier<? extends Throwable> causeSupplier) {
+        requireNonNull(causeSupplier, "causeSupplier");
+        abort0(causeSupplier);
+    }
+
+    private void abort0(Supplier<? extends Throwable> causeSupplier) {
         final SubscriptionImpl currentSubscription = subscription;
         if (currentSubscription != null) {
-            cancelOrAbort(false);
+            cancelOrAbort(false, causeSupplier);
             return;
         }
-
-        final SubscriptionImpl newSubscription = new SubscriptionImpl(
-                this, AbortingSubscriber.get(), ImmediateEventExecutor.INSTANCE, false, false);
-        if (subscriptionUpdater.compareAndSet(this, null, newSubscription)) {
-            // We don't need to invoke onSubscribe() for AbortingSubscriber because it's just a placeholder.
-            invokedOnSubscribe = true;
-        }
-        cancelOrAbort(false);
+        cancelOrAbort(false, causeSupplier);
```

Comment:
If you move this subscription logic inside `if (setState(OPEN, CLEANUP)`, it will never be executed if a user closed the stream before a subscription is made, no?

```java
DefaultStreamMessage m = new DefaultStreamMessage();
m.write(...);
m.close();
m.abort();
m.subscribe(...); // Will succeed, but must fail.
```

Reply:
Please add a test case for this.